### PR TITLE
Allow reading data files bundled within theme-gems

### PR DESCRIPTION
**Pull request by ashmaroli**
If a theme gem contains a `_data` folder at its root, its files will be read and added to the site's internal data hash. The theme-gem's data hash will be overridden by data at source if the same 'key-value' pair exists there.
## 

TODO:
- [x] Add tests
- [x] Add cucumber feature
- [x] Add Documentation
## 

Ref: #5434 
/cc @jekyll/ecosystem, @envygeeks, @pathawks

### CODE REVIEW

**Comment by pathawks** [comment]
For what purpose? Seems like a violation of separation of design from content.

I feel like there must be a better way to do i18n.

**Comment by ashmaroli** [comment]
> For what purpose?

Its not just for i18n _per se_.. but having this can lead to creating a more functional theme-gem.. I dont have a solid use case for this as of now.. will be fleshed out it in the referenced ticket (no. 5434)

--
> Seems like a violation of separation of design from content.

No.. not all.. 
though my original intention was to use it for structural elements like define a navigation system using `navigation.yml` , etc.. I now see the possibilities for built-in i18n as well..

--
P.S.  Will be away for a few days. Feedback welcome.

**Comment by benbalter** [comment]
@ashmaroli thanks for opening up this PR. I really like the vision for themes that you laid out in #5434. 

> I dont have a solid use case for this as of now.

Before we move forward, I'd feel a lot more comfortable if we understood *why* we were building the feature, rather than just building it for a sense of symetry with support for other Jekyll-specific folders, or because it's easy to do so. As was mentioned over in https://github.com/jekyll/jekyll/issues/5350, I share @pathawks concerns that themes should only know about presentation, not content (to ensure theme fungability).

We really, really need to figure out a solid I18N implemntation, but I don't know that `_data` folder support for themes is it.

**Comment by spudowiar** [comment]
> No.. not all..
> though my original intention was to use it for structural elements like define a navigation system using navigation.yml , etc.. I now see the possibilities for built-in i18n as well..

Seems like a good use case... for when you are using `<site root>/_data/navigation.yml` (which already works), **not** `<theme root>/_data/navigation.yml` :smile:

**Comment by ashmaroli** [comment]
> I share Pat Hawks' concerns that themes should only know about presentation, not content (to ensure theme fungibility).

I care about theme fungibility too and this feature does not mess with content in the manner you guys fear. Content, _to me,_ comes from markdown files at source, **supported** by files in an _optional_ `_data` folder.

**Comment by oncleben31** [comment]
> I feel like there must be a better way to do i18n.

> We really, really need to figure out a solid I18N implemntation, 

Have you some guideline as I'm trying to add basic i18n features to minima template (in https://github.com/jekyll/minima/pull/60) ?
 Currently I feel the best option is to use locale string file in  `_data` folder. But I'm new to Jekyll so if you have better ideas please share it.

**Comment by dummied** [comment]
Did #5491 (and ashmaroli/jekyll-data) render this moot?

**Comment by ashmaroli** [comment]
@dummied This is still active, provided there's a demand for this to be included in the core

**Comment by Swiftrien** [comment]
I have a theme in which I provide default values for configuration and theme-translation via the _data directory.
If not in the _data directory, how else could I provide default values?

**Comment by ashmaroli** [comment]
@Swiftrien you may also give the [jekyll-data gem](https://github.com/ashmaroli/jekyll-data) a try in the meantime.. 
It provides the same functionality as this pull, and a li'l bit more.. :smiley:

**Comment by Swiftrien** [comment]
Thanks ashmaroli

However that would mean that every user of my theme also has to install it. Thus this is not the answer to this issue..

**Comment by ashmaroli** [comment]
> However that would mean that every user of my theme also has to install it.

@Swiftrien I'm not trying to promote a plugin I authored, but if you add `jekyll-data` as a `runtime_dependency` to your theme's gemspec, it will be automatically installed with your theme-gem.
Its entirely your call. I won't mind if you decide otherwise.

But I'd definitely like it if you gave it a try, and provide a feedback.
I could then update this pull based on that.

Either way, Happy Jekylling :smiley:

**Comment by Swiftrien** [comment]
I have nothing against promotion :-)
[classic-jekyll-theme](https://rubygems.org/gems/classic-jekyll-theme) :-)

I just added the jekyll-data (0.3.0) gem to a website of mine, removed the _data directory, rebuild the site, and everything is as it should be. Thus jekyll-data seems to work fine. Note that I did not use the "theme configuration" only the "regular data" files.

My bigger issue with external contributions is that is takes away my control. And I have had several experiences where external dependencies costs me huge amounts of time. So in general I try to avoid them where possible. As it is, I think that the jekyll-data gem adds enough value that I will probably include it in the next release. (Mainly because I hope it will be part of Jekyll 3.4 anyhow... ;-))

**Comment by yatil** [comment]
I support this PR for the reasons outlined in #6684. 

The actual practical example for us at W3C/WAI is that we have several independent resources that are then sub-moduled into the big site (W3C/wai-website). Every resource can access information in the data files (minor translations but mostly at this stage WCAG 2.0 Success Criteria and techniques). Moving those source files to the theme gives us the possibility to make the changes once and re-use the data in every resource.

**Comment by mattr-** [comment]
This looks ok to me at first glance. I'd be happy to take a deeper look once this up to date with master and the conflicts are resolved.

**Comment by parkr** [comment]
The purpose of themes and gem-based themes especially is to make the look-and-feel of websites portable. Can you help me understand how data files, which I usually see used for content, fall under the category of look-and-feel?

**Comment by DirtyF** [comment]
@parkr Data files are often use to store config options, that would allow theme authors to embed those options with their thems instead of asking users to copy a sample `_config.yml`. Right now gem-based themes can only ship layouts and assets and often that's not all it takes to have a working theme out of the box, e.g. @mmistakes: https://github.com/mmistakes/minimal-mistakes/tree/master/_data

**Comment by Balancingrock** [comment]
In the classic-jekyll-theme I use the data files for:

- Configuration (as DirtyF commented)
- Translations (User can provide their own translations for titles and headers elements)
- Data for site elements.

The last one is indeed content data, but the theme will process this data to build the user's site. That imo is definitely a theme-job, but needs user generated data.

**Comment by yatil** [comment]
- Also example data: Here is an example navigation that you can copy to your data directory and edit. 
- Text descriptions for page elements instead of having them in the template or in the _config.yml

**Comment by ellemenno** [comment]
@parkr my look and feel use case is icons: I'm working on a theme that embeds icons as svg, rather than using an icon font.

I want to provide the theme icon shape paths in the _data folder, and provide an _include macro for the theme and the user to embed icons wherever needed. By retrieving icon data on demand at build time, I can avoid forcing the inclusion of all icons in the generated page html, which saves significant kb.

**Comment by Ryuno-Ki** [comment]
@ellemenno As @mattr- wrote in https://github.com/jekyll/jekyll/pull/5470#issuecomment-358536052

Failing builds + conflicts in the first place :-)

**Comment by ashmaroli** [comment]
> Scheduling this for inclusion in 4.1

That's awesome news @mattr-. Thank you.

File: lib/jekyll/reader.rb [hunk 1]
```diff
@@ -183,5 +183,21 @@ def read_included_file(entry_path)
         site.static_files.concat(StaticFileReader.new(site, dir).read(file))
       end
     end
+
+    # Read data files within current site.
+    #
+    # If the site uses a theme and has a `_data` directory at its root, read the files within
```
Comment (mattr-) [inline lib/jekyll/reader.rb:189]:
```suggestion
    # If the site uses a theme and the theme has a `_data` directory at its root, read the files within
```

**Comment by DirtyF** [comment]
Waiting for the latest tests to pass to merge...

**Comment by ashmaroli** [comment]
In light of recently acquired knowledge that unlike layouts or includes which are overridden entirely by the namesake at source, data read-in may at times persist partially warranting additional user-intervention to override the persisting  data, I'm withdrawing this submission.

Those interested in this functionality may opt to use the [`jekyll-data`](https://rubygems.org/gems/jekyll-data) plugin instead.

**Comment by ashmaroli** [comment]
@yatil While I understand the sentiment, please consider the following scenario:
A theme has bundled the following data file named `social_links.yml`:
```yaml
twitter: jekyllrb
github:  jekyll
```
And the theme uses the following HTML *include* to render social media icons:
```html
{%- assign social = site.data.social_links -%}
<ul class="social-media-list">
  {% if social.github %}<li><a href="https://github.com/{{ social.github }}" title="{{ social.github  }}"><svg class="svg-icon"><use xlink:href="{{ '/assets/imgs/social-icons.svg#github' | relative_url }}"></use></svg></a></li>{% endif %}
  {% if social.twitter %}<li><a href="https://twitter.com/{{ social.twitter }}" title="{{ social.twitter  }}"><svg class="svg-icon"><use xlink:href="{{ '/assets/imgs/social-icons.svg#twitter' | relative_url }}"></use></svg></a></li>{%- endif -%}
{% if social.linkedin %}<li><a  href="https://www.linkedin.com/in/{{ social.linkedin  }}" title="{{ social.linkedin }}"><svg class="svg-icon"><use xlink:href="{{ '/assets/imgs/social-icons.svg#linkedin' | relative_url }}"></use></svg></a></li>{% endif %}
  <!-- ... so on ... -->
</ul>
```
Now, suppose one of the theme's user doesn't have a GitHub account or a Twitter handle.. they think that all they need to do is override the theme's data file with one at their source:
```yaml
# source/_data/social_links.yml

linkedin: johndoe
```

*However, when they build their site, all three social icons are rendered by default.*
The user has to then explicitly set `github` and `twitter` keys to `null` to not render them, A couple of keys may not be a hassle. But imagine if the theme supports over 50 social accounts.., the user would have to manually disable over 4 dozens of keys...........

**Comment by yatil** [comment]
@ashmaroli 

> However, when they build their site, all three social icons are rendered by default.

Because both, the theme's `social_links.yml` and the site’s `social_links.yml` are interpreted. But, theoretically, wouldn’t it be possible to say “if the site has a social_links.yml, ignore the theme’s social_links.yml”?

I personally can see both behaviors to be beneficial, but one would then need to be able to configure it and that might create problems.

Thanks for exploring the idea, it is much appreciated.

**Comment by ashmaroli** [comment]
> would then need to be able to configure it and that might create problems

Jekyll 4.0 introduced the ability to import *configuration* from a theme-gem and merge it with the one from the source. It led to problems similar to what I described above.
And now Jekyll 4.1 will make amends by providing users with an option to disable loading the configuration from the theme.

> theoretically, wouldn’t it be possible to say “if the site has a social_links.yml, ignore the theme’s social_links.yml”?

Yes, that is possible. I had considered that but there's catch though.
`_data/foo.yml` containing top-level key `bar:` and `_data/foo/bar.yml` are equivalent and it is not  possible to know about *top-level keys* without parsing the data file first.

**Pull request closed without merging**